File: client_packages/WiredPlayers/business/hairdresser.js
const maleFaceOptions = [
	{'desc': 'hairdresser.hair', 'minValue': 0, 'maxValue': 36}, {'desc': 'hairdresser.hair-primary', 'minValue': 0, 'maxValue': 63}, 
	{'desc': 'hairdresser.hair-secondary', 'minValue': 0, 'maxValue': 63}, {'desc': 'hairdresser.eyebrows', 'minValue': 0, 'maxValue': 33}, 
	{'desc': 'hairdresser.eyebrows-color', 'minValue': 0, 'maxValue': 63}, {'desc': 'hairdresser.beard', 'minValue': -1, 'maxValue': 36}, 
	{'desc': 'hairdresser.beard-color', 'minValue': 0, 'maxValue': 63}
];

const femaleFaceOptions = [
	{'desc': 'hairdresser.hair', 'minValue': 0, 'maxValue': 38}, {'desc': 'hairdresser.hair-primary', 'minValue': 0, 'maxValue': 63}, 
	{'desc': 'hairdresser.hair-secondary', 'minValue': 0, 'maxValue': 63}, {'desc': 'hairdresser.eyebrows', 'minValue': 0, 'maxValue': 33}, 
	{'desc': 'hairdresser.eyebrows-color', 'minValue': 0, 'maxValue': 63}
];

let faceHairArray = [];

mp.events.add('showHairdresserMenu', (businessName) => {
	// Get the player and its sex
	let player = mp.players.local;
	let sex = player.getVariable('PLAYER_SEX');
	
	// Add the options
	let faceOptions = JSON.stringify(sex === 0 ? maleFaceOptions : femaleFaceOptions);
	
	// Inicializamos los valores
	faceHairArray.push(player.getVariable('HAIR_MODEL'));
	faceHairArray.push(player.getVariable('FIRST_HAIR_COLOR'));
	faceHairArray.push(player.getVariable('SECOND_HAIR_COLOR'));
	faceHairArray.push(player.getVariable('EYEBROWS_MODEL'));
	faceHairArray.push(player.getVariable('EYEBROWS_COLOR'));
	faceHairArray.push(player.getVariable('BEARD_MODEL'));
	faceHairArray.push(player.getVariable('BEARD_COLOR'));
	
	// Create hairdressers' menu
	mp.events.call('createBrowser', ['package://WiredPlayers/statics/html/sideMenu.html', 'populateHairdresserMenu', faceOptions, JSON.stringify(faceHairArray), businessName]);
});

mp.events.add('updateFacialHair', (slot, value) => {
	// Get the player
	let player = mp.players.local;
	
	// Save the new value
	faceHairArray[slot] = value;
	
	// Update the player's head
	player.setComponentVariation(2, faceHairArray[0], 0, 0);
	player.setHairColor(faceHairArray[1], faceHairArray[2]);
	player.setHeadOverlay(1, faceHairArray[5], 0.99, faceHairArray[6], 0);
	player.setHeadOverlay(2, faceHairArray[3], 0.99, faceHairArray[4], 0);
});

mp.events.add('applyHairdresserChanges', () => {
	let generatedFace = {};
	
	generatedFace.hairModel = faceHairArray[0];
	generatedFace.firstHairColor = faceHairArray[1];
	generatedFace.secondHairColor = faceHairArray[2];
	generatedFace.eyebrowsModel = faceHairArray[3];
	generatedFace.eyebrowsColor = faceHairArray[4];
	generatedFace.beardModel = faceHairArray[5];
	generatedFace.beardColor = faceHairArray[6];

	// Apply the changes to the player
	mp.events.callRemote('changeHairStyle', JSON.stringify(generatedFace));
});

mp.events.add('cancelHairdresserChanges', () => {
	// Get the player
	let player = mp.players.local;

	// Load the variables
	let hairModel = player.getVariable('HAIR_MODEL');
	let firstHairColor = player.getVariable('FIRST_HAIR_COLOR');
	let secondHairColor = player.getVariable('SECOND_HAIR_COLOR');
	let eyebrowsModel = player.getVariable('EYEBROWS_MODEL');
	let eyebrowsColor = player.getVariable('EYEBROWS_COLOR');
	let beardModel = player.getVariable('BEARD_MODEL');
	let beardColor = player.getVariable('BEARD_COLOR');

	// Revert the changes
	player.setComponentVariation(2, hairModel, 0, 0);
	player.setHairColor(firstHairColor, secondHairColor);
	player.setHeadOverlay(1, beardModel, 0.99, beardColor, 0);
	player.setHeadOverlay(2, eyebrowsModel, 0.99, eyebrowsColor, 0);
});

File: XenRP-Client/business/Hairdresser.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using RAGE;
using RAGE.Elements;
using XenRP.Client.globals;
using XenRP.Client.model;

namespace XenRP.Client.business {
    internal class Hairdresser : Events.Script {
        private List<int> facialHair;
        private FacialHair initialHair;

        public Hairdresser() {
            Events.Add("showHairdresserMenu", ShowHairdresserMenuEvent);
            Events.Add("updateFacialHair", UpdateFacialHairEvent);
            Events.Add("applyHairdresserChanges", ApplyHairdresserChangesEvent);
            Events.Add("cancelHairdresserChanges", CancelHairdresserChangesEvent);
        }

        private void ShowHairdresserMenuEvent(object[] args) {
            // Get the variables from the arguments
            var sex = Convert.ToInt32(args[0]);
            var skinJson = args[1].ToString();
            var businessName = args[2].ToString();

            // Add the options
            var faceOption = JsonConvert.SerializeObject(sex == Constants.SEX_MALE
                ? Constants.MALE_FACE_OPTIONS
                : Constants.FEMALE_FACE_OPTIONS);

            // Initialize the face values
            initialHair = JsonConvert.DeserializeObject<FacialHair>(skinJson);

            facialHair = new List<int>();
            facialHair.Add(initialHair.hairModel);
            facialHair.Add(initialHair.firstHairColor);
            facialHair.Add(initialHair.secondHairColor);
            facialHair.Add(initialHair.eyebrowsModel);
            facialHair.Add(initialHair.eyebrowsColor);
            facialHair.Add(initialHair.beardModel);
            facialHair.Add(initialHair.beardColor);

            // Create hairdressers' menu
            Browser.CreateBrowserEvent(new object[] {
                "package://statics/html/sideMenu.html", "populateHairdresserMenu", faceOption,
                JsonConvert.SerializeObject(facialHair), businessName
            });
        }

        private void UpdateFacialHairEvent(object[] args) {
            // Get the variables from the arguments
            var slot = Convert.ToInt32(args[0]);
            var value = Convert.ToInt32(args[1]);

            // Save the new value
            facialHair[slot] = value;

            // Check if the beard is out of range
            var beardModel = facialHair[5] < 0 ? 255 : facialHair[5];

            // Update the player's head
            Player.LocalPlayer.SetComponentVariation(2, facialHair[0], 0, 0);
            Player.LocalPlayer.SetHairColor(facialHair[1], facialHair[2]);
            Player.LocalPlayer.SetHeadOverlay(1, beardModel, 1.0f);
            Player.LocalPlayer.SetHeadOverlay(2, facialHair[3], 1.0f);
            Player.LocalPlayer.SetHeadOverlayColor(1, 1, facialHair[6], 0);
            Player.LocalPlayer.SetHeadOverlayColor(2, 1, facialHair[4], 0);
        }

        private void ApplyHairdresserChangesEvent(object[] args) {
            var generatedFace = new FacialHair();
            generatedFace.hairModel = facialHair[0];
            generatedFace.firstHairColor = facialHair[1];
            generatedFace.secondHairColor = facialHair[2];
            generatedFace.eyebrowsModel = facialHair[3];
            generatedFace.eyebrowsColor = facialHair[4];
            generatedFace.beardModel = facialHair[5] < 0 ? 255 : facialHair[5];
            generatedFace.beardColor = facialHair[6];

            Events.CallRemote("changeHairStyle", JsonConvert.SerializeObject(generatedFace));
        }

        private void CancelHairdresserChangesEvent(object[] args) {
            // Revert the changes
            Player.LocalPlayer.SetComponentVariation(2, initialHair.hairModel, 0, 0);
            Player.LocalPlayer.SetHairColor(initialHair.firstHairColor, initialHair.secondHairColor);
            Player.LocalPlayer.SetHeadOverlay(1, initialHair.beardModel, 1.0f);
            Player.LocalPlayer.SetHeadOverlay(2, initialHair.eyebrowsModel, 1.0f);
            Player.LocalPlayer.SetHeadOverlayColor(1, 1, initialHair.beardColor, 0);
            Player.LocalPlayer.SetHeadOverlayColor(2, 1, initialHair.eyebrowsColor, 0);
        }
    }
}